vm/parser.ts

```diff
@@ -711,6 +861,17 @@ module J2ME {
       }
     }
 
+    getClassNameSlow(): string {
+      return this._name || (this._name = ByteStream.readString(this.utf8Name));
+    }
```

Comment:
> I tried a quick benchmark, and the getXXX methods are the fastest, probably because they get inlined and have monomorphic call sites:

Erm, but your benchmark doesn't measure what your code actually does, nor what I suggest it should do instead. Here's a [more accurate one in JSFiddle](http://jsfiddle.net/dt4z21af/) based on the actual JavaScript code to which the TypeScript compiles. It shows that the memoizer is faster.

Part of the problem may be that your version mutates the shape of the object, since it doesn't define *this._name* in the *ClassInfo* constructor. And [fixing that in an updated fiddle](http://jsfiddle.net/dt4z21af/1/) brings it in line with the memoizer's perf (and works as long as you ensure that the get*() method correctly distinguishes between the initial value and the eventual one; which can probably be a simple check for truthiness), for 1000 objects with 1000 accesses each, anyway.

But bump those numbers, and your version starts slowing down again.

Here's [one thousand objects with ten thousand accesses](http://jsfiddle.net/dt4z21af/2/). And [ten thousand objects with one thousand accesses](http://jsfiddle.net/dt4z21af/3/). I'm not sure what these numbers will actually look like in the field, but I see ~4500 calls to *getClassNameSlow* on startup when loading the very simple [InitMidlet](https://github.com/mozilla/j2me.js/blob/master/tests/midlets/InitMidlet.java).

> To keep the size of the patch down, I used getters for name and signature so I wouldn't have to update all the call sites.

That's a worthy concern, but this changeset would be even smaller if you didn't replace all the *className* references with *getClassNameSlow* calls.